### PR TITLE
Cooperative __init__ for Qt4 canvas.

### PR DESCRIPTION
**Pull request by anntzer**
Alternate for #9048 (basically pulling out the relevant parts of #8771, but using a local monkey-patch approach.
I personally find this approach clearer in intent ("PyQt4 does not support multiple inheritance so let's just locally patch it so that it does") than #9048.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backends/backend_qt5.py [hunk 1]
```diff
@@ -132,6 +133,45 @@ def _create_qApp():
...
+            mro = type(self).__mro__
+            next_coop_init = next(
+                cls for cls in mro[mro.index(QtWidgets.QWidget) + 1:]
+                if cls.__module__.split(".")[0] not in [
+                    "PyQt4", "sip", "PySide", "PySide2", "Shiboken"])
+            next_coop_init.__init__(self, *args, **kwargs)
+
+        @functools.wraps(__init__)
+        def wrapper(self, **kwargs):
+            QtWidgets.QWidget.__init__ = cooperative_qwidget_init
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_qt5.py]: This needs to be in a `try..finally` block.

**Comment by tacaswell** [comment]
I lean towards this as the better solution (I started writing a comment about how I was even between them and in the process convinced my self to prefer this one).

This has the advantage of being cleaner and more localized but hard-codes the non-cooperative modules in.  However, odds of another binding showing up are low and doing the sub-classing and replacing `__init__` while waiting for changes to be merged upstream isn't that bad.

#9048 uses 'simpler' python, but it makes the class zoo one class bigger and has to use two differently named 'magic' cut-out function, and is spread across a couple of files.   It makes the unlikely extension case easier (once you understand the code), but harder to understand to begin with.

File: lib/matplotlib/backends/backend_qt5.py (hunk 1 above)
Comment (anntzer) [inline lib/matplotlib/backends/backend_qt5.py]: done

**Comment by anntzer** [comment]
try... finally added.
Honestly, I think anyone coming up with a new qt binding that does not support cooperative inheritance does not deserve being supported by matplotlib :-) (you can make the mistake once, but you should learn from previous mistakes!)

File: lib/matplotlib/backends/backend_qt5.py (hunk 1 above)
Comment (tacaswell) [inline lib/matplotlib/backends/backend_qt5.py]: The initial monkey-patch should also be in the try block... `ctrl-c` can happen anywhere.

Comment (anntzer) [inline lib/matplotlib/backends/backend_qt5.py]: Wow, I'm a bit ashamed I never thought about this point :-)

**Comment by efiring** [comment]
Appveyor failure is just a hang at the start of the conda-build on 3.6.

**Pull request merged**